docs/varnish.md

```diff
@@ -0,0 +1,90 @@
+  if (req.http.Purge-Cache-Tags) {
+	ban("obj.http.Purge-Cache-Tags ~ " + req.http.Purge-Cache-Tags);
+  }
+  else {
+	return (synth(403, "Purge-Cache-Tags header missing."));
+  }
+
+  # Throw a synthetic page so the request won't go to the backend.
+  return (synth(200, "Ban added."));
+}
+```
+The following script will accept "BAN" commands from Drupal 8 Purge module and process it accordingly.
+
+After these changes varnish can be restarted and it's ready to accept cache invalidation requests from Drupal 8 Purge module.
+
+
+#### Install Drupal 8 Purge modules
+
+Purge module provides functionality to expose cache tags in the header of the response. Varnish by default will keep header saved for every cache entry and that information will be used later to invalidate cache entries. You can download purge from https://www.drupal.org/project/purge and additional purge_purger_http is required. It can be downloaded from https://www.drupal.org/project/purge_purger_http. To download them over drush simple execute the following command:
+```bash
+drush dl purge, purge_purger_http
+```
+
+After that enable following modules:
+- Purge (purge) - base Purge module
+- Purge Tokens (purge_tokens) - required to replace generic cache tag tokens
+- Purge UI (purge_ui) - user interface for Purge configuration pages
+- Late runtime processor (purge_processor_lateruntime) - purge process, it will trigger cache invalidation on any core cache invalidation (fe. article save, media entity save, etc.)
+- Core tags queuer (purge_queuer_coretags) - provides queue core cache tag invalidation
+- Generic HTTP Purger (purge_purger_http) - makes BAN request, to execute cache invalidation for Varnish
+- Generic HTTP Tags Header (purge_purger_http_tagsheader) - exposes required header for Varnish
+
+To enable modules over drush, execute following command:
+```bash
+drush en purge, purge_tokens, purge_ui, purge_processor_lateruntime, purge_queuer_coretags, purge_purger_http, purge_purger_http_tagsheader
+```
+
+After these modules are enabled, Drupal should provide the ```Purge-Cache-Tags``` header. That header property contains all cache tags for the loaded page.
+
+#### Setup Drupal site to use Purge
+
+On the Drupal 8 site open: Configuration -> Development -> Performance page (```admin/config/development/performance```). Enable caching and set it really high, ideally max period (1 year). Save configuration and after that open Purge configuration page (```admin/config/development/performance/purge```).
+
+On that page do following configuration:
+1. Click "Add Purger"
+2. Choose "HTTP Bundled Purger"
+3. Click "Add"
+4. HTTP Bundled Purger will be added with generic name
+5. Click drop-down button and choose "Configuration"
+6. Set "Name" for Purger (fe. Varnish Bundled HTTP Purger)
+7. Adjust Hostname and Port to match your Varnish server
+8. Click "Headers"
+9. Create Header - Name: ```Purge-Cache-Tags``` - Value: ```[invalidations:separated_pipe]```
```

Comment:
This header should not be exposed to the web browser. Varnish should remove it from the response before sending it to the browser. Furthermore it might be possible that the default configuration of varnish might run into problems, when this header gets to big. Please investigate this and document possible configuration options, that might help to mitigate the problem.